core/src/main/java/org/elasticsearch/action/search/RemoteClusterService.java

```diff
@@ -0,0 +1,395 @@
+                    remoteClusters.put(entry.getKey(), remote);
+                }
+
+                // now update the seed nodes no matter if it's new or already existing
+                RemoteClusterConnection finalRemote = remote;
+                remote.updateSeedNodes(entry.getValue(), ActionListener.wrap(
+                    response -> {
+                        if (countDown.countDown()) {
+                            connectionListener.onResponse(response);
+                        }
+                    },
+                    exception -> {
+                        if (countDown.fastForward()) {
+                            connectionListener.onFailure(exception);
+                        }
+                        if (finalRemote.isClosed() == false) {
+                            logger.warn("failed to update seed list for cluster: " + entry.getKey(), exception);
+                        }
+                    }));
+            }
+        }
+        this.remoteClusters = Collections.unmodifiableMap(remoteClusters);
+    }
+
+    /**
+     * Returns <code>true</code> if at least one remote cluster is configured
+     */
+    boolean isCrossClusterSearchEnabled() {
+        return remoteClusters.isEmpty() == false;
+    }
+
+    /**
+     * Filters out indices that refer to a remote cluster and adds them to the given per cluster indices map.
+     *
+     * @param perClusterIndices a map to fill with remote cluster indices from the given request indices
+     * @param requestIndices the indices in the search request to filter
+     * @param indexExists a predicate that can test if a certain index or alias exists
+     *
+     * @return all indices in the requestIndices array that are not remote cluster indices
+     */
+    public String[] filterIndices(Map<String, List<String>> perClusterIndices, String[] requestIndices, Predicate<String> indexExists) {
+        List<String> localIndicesList = new ArrayList<>();
+        for (String index : requestIndices) {
+            int i = index.indexOf(REMOTE_CLUSTER_INDEX_SEPARATOR);
+            if (i >= 0) {
+                String remoteCluster = index.substring(0, i);
+                if (isRemoteClusterRegistered(remoteCluster)) {
+                    if (indexExists.test(index)) {
+                        // we use : as a separator for remote clusters. might conflict if there is an index that is actually named
+                        // remote_cluster_alias:index_name - for this case we fail the request. the user can easily change the cluster alias
+                        // if that happens
+                        throw new IllegalArgumentException("Index " + index + " exists but there is also a remote cluster named: "
+                            + remoteCluster + " can't filter indices");
+                    }
+                    String remoteIndex = index.substring(i + 1);
+                    List<String> indices = perClusterIndices.get(remoteCluster);
```

Comment:
If you do not go with my [suggestion](https://github.com/elastic/elasticsearch/pull/22502/files#r96339204), this can be simplified to:

```java
perClusterIndices.computeIfAbsent(remoteCluster, k -> new ArrayList<>()).add(remoteIndex);
```